plonky2x/src/frontend/vars/variable.rs

```diff
@@ -107,3 +109,142 @@ impl<L: PlonkParameters<D>, const D: usize> One<L, D> for Variable {
         Variable(builder.api.one())
     }
 }
+
+impl<L: PlonkParameters<D>, const D: usize> LessThan<L, D> for Variable {
+    type Output = BoolVariable;
+
+    fn lt(self, rhs: Variable, builder: &mut CircuitBuilder<L, D>) -> Self::Output {
+        // FIXME: Check for overflows and edge cases
+        const NUM_LIMBS: usize = 64;
+        let max = builder.constant(L::Field::from_canonical_u64(1 << (NUM_LIMBS - 1)));
+        let _one: Variable = builder.constant(L::Field::from_canonical_u64(1));
+        // from circomlib: https://github.com/iden3/circomlib/blob/master/circuits/comparators.circom#L89
+        let tmp = builder.add(self, max);
+        let res = builder.sub(tmp, rhs);
+        // typically stored in BE, but we can just index instead of reversing
+        let res_bits = builder.api.split_le(res.0, NUM_LIMBS);
+        let last_bit = Variable::from(res_bits[NUM_LIMBS - 1].target);
+        let lt_var = builder.sub(_one, last_bit);
+        BoolVariable::from(lt_var)
+    }
+}
+
+impl<L: PlonkParameters<D>, const D: usize> LessThanOrEqual<L, D> for Variable {
+    type Output = BoolVariable;
+
+    fn lte(self, rhs: Variable, builder: &mut CircuitBuilder<L, D>) -> Self::Output {
+        let lt = self.lt(rhs, builder);
+        let eq = builder.is_equal(self, rhs);
+        builder.or(lt, eq)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::backend::circuit::DefaultParameters;
+    use crate::backend::circuit::config::PlonkParameters;
+    use crate::prelude::*;
+
+    type L = DefaultParameters;
+    const D: usize = 2;
+
+    #[test]
+    fn test_lt() {
+        let mut builder = CircuitBuilder::<L, D>::new();
+        let lhs = builder.read::<Variable>();
+        let rhs = builder.read::<Variable>();
+        let less_than = builder.lt(lhs, rhs);
+        builder.write(less_than);
+
+        let circuit = builder.build();
+
+        let rand_lhs = rand::random::<u64>();
```

Comment:
Won't this fail if `rand_lhs` is > the field order? or there are some edge cases here I think that we shoudl be careful of

Reply:
fixed in [c5c901d](https://github.com/succinctlabs/succinctx/pull/152/commits/c5c901d1da6ce4cb7d2237f81266d01e054632d0)